sklearn/cluster/optics_.py

```diff
@@ -527,3 +565,352 @@ def cluster_optics_dbscan(reachability, core_distances, ordering, eps=0.5):
+
+                # line (**), sc2*
+                if reachability_plot[c_end + 1] * xi_complement < D['mib']:
+                    continue
+
+                # Definition 11: criterion 4
+                if D_max * xi_complement >= reachability_plot[c_end + 1]:
+                    # Find the first index from the left side which is almost
+                    # at the same level as the end of the detected cluster.
+                    while (reachability_plot[c_start + 1] >
+                           reachability_plot[c_end + 1]
+                           and c_start < c_end):
+                        c_start += 1
+                elif reachability_plot[c_end + 1] * xi_complement >= D_max:
+                    # Find the first index from the right side which is almost
+                    # at the same level as the beginning of the detected
+                    # cluster.
+                    while (reachability_plot[c_end] > D_max
+                           and c_end > c_start):
+                        c_end -= 1
+
+                # predecessor correction
+                if predecessor_correction:
+                    c_start, c_end = _correct_predecessor(reachability_plot,
+                                                          predecessor,
+                                                          c_start,
+                                                          c_end)
+                if c_start is None:
+                    continue
+
+                # Definition 11: criterion 3.a
+                if c_end - c_start + 1 < min_cluster_size:
+                    continue
+
+                # Definition 11: criterion 1
+                if c_start > c_end:
+                    continue
+
+                # Definition 11: criterion 2
+                if c_end < c_start:
+                    continue
+
+                U_clusters.append((c_start, c_end))
+
+            # add smaller clusters first.
+            U_clusters.reverse()
+            clusters.extend(U_clusters)
+
+            if U['end'] == n_samples - 1:
+                break
+
+    return np.array(clusters)
+
+
+def _extract_xi_labels(ordering, clusters):
+    """Extracts the labels from the clusters returned by `_xi_cluster`.
+    We rely on the fact that clusters are stored
+    with the smaller clusters coming before the larger ones.
+
+    Parameters
+    ----------
+    ordering : array, shape (n_samples)
+        The ordering of points calculated by OPTICS
+
+    clusters : array, shape (n_clusters, 2)
+        List of clusters i.e. (start, end) tuples,
+        as returned by `_xi_cluster`.
+
+    Returns
+    -------
+    labels : array, shape (n_samples)
+    """
+
+    labels = np.full(len(ordering), -1, dtype=int)
+    label = 0
+    for c in clusters:
+        if not np.any(labels[c[0]:(c[1] + 1)] != -1):
+            labels[c[0]:(c[1] + 1)] = label
+            label += 1
+    labels[ordering] = labels.copy()
```

Comment:
do not need to copy?

Reply:
The copy is there because

``` python
import numpy as np

ordering = np.arange(0, 10)
np.random.seed(42)
np.random.shuffle(ordering)

clusters = np.array([[0, 3], [0, 5], [7, 8], [6, 9]])

labels = np.full(len(ordering), -1, dtype=int)
label = 0
for c in clusters:
    if not np.any(labels[c[0]:(c[1] + 1)] != -1):
        labels[c[0]:(c[1] + 1)] = label
        label += 1
labels[ordering] = labels.copy()
print(labels)

labels = np.full(len(ordering), -1, dtype=int)
label = 0
for c in clusters:
    if not np.any(labels[c[0]:(c[1] + 1)] != -1):
        labels[c[0]:(c[1] + 1)] = label
        label += 1
labels[ordering] = labels
print(labels)
```
produces

```
[ 0  0 -1  1  1  0 -1 -1  0 -1]
[ 0  0  0  0 -1  0 -1 -1  0 -1]
```
The two outputs are not the same.